Objects/setobject.c

```diff
@@ -62,6 +62,152 @@ static PyObject _dummy_struct;
 
 #define dummy (&_dummy_struct)
 
+#define SET_LOOKKEY_FOUND 1
+#define SET_LOOKKEY_NO_MATCH 0
+#define SET_LOOKKEY_ERROR (-1)
+#define SET_LOOKKEY_CHANGED (-2)
+#define SET_LOOKKEY_EMPTY (-3)
+
+#ifdef Py_GIL_DISABLED
+
+#define LOAD_HASH(entry) _Py_atomic_load_ssize_acquire(&(entry)->hash)
```

Comment:
I think LOAD_HASH and STORE_HASH can use relaxed ordering if we make sure they are stored before STORE_KEY with release order so that it would synchronize with LOAD_KEY with acquire ordering.

Reply:
Yeah, that seems to be okay.  At least, TSAN doesn't warn about it.  I've switched the code to directly use the `FT_ATOMIC_*` macros since I think that's clearer (even if it's more verbose).